File: front/angular-front/src/app/models/domainModel/englishName/english-name.ts
import {PreferNominal} from "../../prefer-nominal";
import * as E from "fp-ts/Either";

export class EnglishName {
  // noinspection JSUnusedGlobalSymbols
  readonly _englishNameBrand: PreferNominal;

  constructor(readonly value: string) {
    if (!EnglishName.isValid(value)) {
      throw new TypeError(this.requirementErrorMessage);
    }
  }

  private get requirementErrorMessage(): string {
    return `EnglishName must between 1 to 100 length and only use alphabet or numerical, but value is "${this.value}"`;
  }

  equals(other: EnglishName): boolean {
    return this.value === other.value;
  }

  static validate(value: string): E.Either<string, EnglishName> {
    if (this.isValid(value)) {
      return E.right(new EnglishName(value));
    }
    return E.left(this.validationErrorMessage);
  }

  static isValid(value: string): boolean {

    console.log(this.mustBe1To100Characters(value));
    console.log(this.mustOnlyAlphanumerical(value));

    return this.mustBe1To100Characters(value) && this.mustOnlyAlphanumerical(value);
  }

  static validationErrorMessage: string = '英語名は1文字以上100文字以下である必要があります';

  private static mustBe1To100Characters(value: string): boolean {
    return value.length >= 1 && value.length <= 100;
  }

  private static mustOnlyAlphanumerical(value: string): boolean {
    const regex = new RegExp('^[a-zA-Z\\d]{1,100}$');
    return regex.test(value);
  }
}

File: sdmt-core/src/main/scala/dev/tchiba/sdmt/core/domainmodel/EnglishName.scala
package dev.tchiba.sdmt.core.domainmodel

import dev.tchiba.arch.ddd.ValueObject

/**
 * ドメインモデルの英語名
 *
 * @param value 1文字以上100文字以下の英語もしくは数字文字列。半角スペースを含まない。
 */
final class EnglishName private (val value: String) extends ValueObject {

  import EnglishName._

  require(requirement(value), requirementErrorMessage(value))

  override def equals(other: Any): Boolean = other match {
    case that: EnglishName => value == that.value
    case _                 => false
  }

  override def hashCode(): Int = 31 * value.##

  override def toString = s"EnglishName($value)"
}

object EnglishName {
  def apply(value: String) = new EnglishName(value)

  def validate(value: String): Either[String, EnglishName] =
    Either.cond(requirement(value), apply(value), requirementErrorMessage(value))

  private def requirementErrorMessage(value: String): String =
    s"EnglishName must between 1 to 100 length and only use alphabet or numerical, but value is $value"

  private def requirement(value: String): Boolean = valueLengthMustBe1To100(value) && mustOnlyAlphanumerical(value)

  private def valueLengthMustBe1To100(value: String): Boolean = value.nonEmpty && value.lengthIs <= 100

  private def mustOnlyAlphanumerical(value: String): Boolean = "[a-zA-Z\\d]{1,100}".r.matches(value)
}
